pkg/querier/remote/read.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/logqlmodel"
+	"github.com/grafana/loki/pkg/querier"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/stores/index/stats"
+)
+
+// ReadConfig is the configuration for reading from remote storage.
+// code from /github.com/prometheus/prometheus/config/config.go:868
+type ReadConfig struct {
+	URL   *config.URL `yaml:"url"`
+	Name  string      `yaml:"name,omitempty"`
+	OrgID string      `yaml:"orgID,omitempty"`
+
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+
+	//todo:@liguozhong support http herder
+	Headers map[string]string `yaml:"headers,omitempty"`
+	//todo:@liguozhong support RemoteTimeout
+	RemoteTimeout time.Duration `yaml:"remote_timeout,omitempty"`
+}
+
+func NewQuerier(name string, remoteReadConfig ReadConfig) (querier.Querier, error) {
+	client := &client.DefaultClient{
+		OrgID:   remoteReadConfig.OrgID,
+		Address: remoteReadConfig.URL.String(),
+		TLSConfig: config.TLSConfig{
+			InsecureSkipVerify: remoteReadConfig.HTTPClientConfig.TLSConfig.InsecureSkipVerify,
+		},
+	}
+	if remoteReadConfig.HTTPClientConfig.BasicAuth != nil {
+		client.Username = remoteReadConfig.HTTPClientConfig.BasicAuth.Username
+		client.Password = strings.TrimSpace(string(remoteReadConfig.HTTPClientConfig.BasicAuth.Password))
+	}
+
+	return &Querier{
+		client: client,
+		name:   name,
+	}, nil
+}
+
+type Querier struct {
+	client client.Client
+	codec  queryrangebase.Codec
+	name   string
+}
+
+func (q Querier) SelectLogs(ctx context.Context, params logql.SelectLogParams) (iter.EntryIterator, error) {
+	response, err := q.client.QueryRange(ctx, params.Selector, int(params.Limit), params.Start, params.End, params.Direction, 0, 0, false)
+	if err != nil {
+		return nil, err
+	}
+	if response.Status != loghttp.QueryStatusSuccess {
+		return nil, errors.Errorf("remote read Querier selectLogs fail,response.Status %v", response.Status)
+	}
+
+	streams, ok := response.Data.Result.(loghttp.Streams)
+	if !ok {
+		return nil, errors.New("remote read Querier selectLogs fail,value cast (loghttp.Streams) fail")
+	}
+	return iter.NewStreamsIterator(streams.ToProto(), params.Direction), nil
+}
+
+func (q Querier) SelectSamples(ctx context.Context, params logql.SelectSampleParams) (iter.SampleIterator, error) {
+	response, err := q.client.QueryRange(ctx, params.Selector, 1, params.Start, params.End, logproto.FORWARD, 0, 0, false)
+	if err != nil {
+		return nil, err
+	}
+	if response.Status != loghttp.QueryStatusSuccess {
+		return nil, errors.Errorf("remote read Querier selectSamples fail,response.Status %v", response.Status)
+	}
+
+	value := response.Data.Result
+	switch value.Type() {
+	case logqlmodel.ValueTypeStreams:
+		return nil, fmt.Errorf("remote read Querier selectLogs fail,Unable to parse unsupported type: %s ", value.Type())
```

Comment:
```suggestion
		return nil, fmt.Errorf("Unable to parse unsupported type: %s ", value.Type())
```

unifies the errors a little.